superduperdb/vector_search/lance.py

```diff
@@ -0,0 +1,178 @@
+    VectorDatabase,
+)
+
+# TODO: Implement a vector index manager for ``lance`` after tidying
+# base classes and interfaces in ``superduperdb.vector_search.base``.
+
+
+class LanceVectorDatabase(VectorDatabase):
+    """A class for managing multiple Lance datasets."""
+
+    def __init__(self, root_dir: str) -> None:
+        """
+        :param root_dir: root directory that contains all Lance datasets
+        """
+
+        self.root_dir = root_dir
+        self._collections: t.Dict[VectorCollectionId, VectorCollection] = {}
+
+    # TODO: ``create`` kwarg is not included in the ABC interface but it is assumed part
+    # of the interface in ``superduperdb/container/vector_index.py``! Short-term fix
+    # is to include it here but long-term fix is to tidy up the ABC interface.
+    def get_table(
+        self, config: VectorCollectionConfig, create: bool = False
+    ) -> VectorCollection:
+        """Retrieve a managed Lance dataset according to the given config."""
+
+        assert config.id.startswith('lance://')
+        uri = config.id.split('lance://')[1]
+        assert uri.startswith(self.root_dir)
+
+        collection = self._collections.get(config.id)
+
+        # TODO: Add support for lance ``version`` parameter
+        if not collection:
+            collection = self._collections[config.id] = LanceVectorCollection(
+                uri=uri, dimensions=config.dimensions
+            )
+        return collection
+
+
+class LanceVectorCollection(VectorCollection):
+    """A class for managing a lance dataset."""
+
+    def __init__(
+        self,
+        uri: str,
+        dimensions: int,
+    ) -> None:
+        """
+        :param uri: URI of the Lance dataset
+        :param dimensions: dimension of the vector embeddings in the Lance dataset
+        """
+        self.uri = uri
+        self.dimensions = dimensions
+
+    @property
+    def _dataset(self) -> lance.LanceDataset:
+        try:
+            return lance.dataset(self.uri)
+        except Exception as e:
+            raise Exception(f'Could not load Lance dataset at {self.uri}') from e
+
+    # TODO: ``upsert`` kwarg is not included in the ABC interface but it is assumed part
+    # of the interface in ``superduperdb/db/mongodb/cdc/vector_task_factory.py``!
+    # Short-term fix is to include it here, long-term fix is to tidy up ABC interface.
+    def add(self, data: t.Sequence[VectorCollectionItem], upsert: bool = False) -> None:
+        """
+        Add vectors to existing Lance dataset or create a new one if it doesn't exist.
+
+        :param data: t.Sequence of ``VectorCollectionItem`` objects.
+        """
+
+        typ = pa.list_(
+            pa.field('values', pa.float32(), nullable=False), self.dimensions
+        )
+        _vecs = pa.array([item.vector for item in data], type=typ)
+        _ids = pa.array([item.id for item in data], type=pa.string())
+        _table = pa.Table.from_arrays([_ids, _vecs], names=['id', 'vector'])
```

Comment:
lets create class attribute for id and vector

names = [self._ID_COLUMN, self._VECTOR_COLUMN]
something like the above